src/CodeGen_RISCV.cpp

```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+    }
+    llvm::Type *llvm_ret_type;
+    if (ret_type.is_vector()) {
+        int lanes = ret_type.lanes();
+        bool scalable = (intrin.ret_type.type_pattern != FixedOrScalableVector::Fixed);
+        if (scalable) {
+            lanes /= effective_vscale;
+        }
+        llvm_ret_type = llvm::VectorType::get(llvm_type_of(ret_type.element_of()),
+                                              lanes, scalable);
+    } else {
+        llvm_ret_type = llvm_type_of(ret_type);
+    }
+
+    llvm_arg_types.push_back(llvm_ret_type);
+    for (const auto &arg_type_pattern : intrin.arg_types) {
+        if (arg_type_pattern.type_pattern == FixedOrScalableVector::Undefined) {
+            break;
+        }
+        Type arg_type = concretize_fixed_or_scalable(arg_type_pattern, bit_width_scale, target.vector_bits);
+
+        bool scalable = (arg_type_pattern.type_pattern != FixedOrScalableVector::Fixed);
+        mangled_name += "." + mangle_vector_argument_type(arg_type, scalable, effective_vscale);
+        llvm::Type *llvm_type;
+        if (arg_type.is_vector()) {
+            int lanes = arg_type.lanes();
+            if (scalable) {
+                lanes /= effective_vscale;
+            }
+            llvm_type = llvm::VectorType::get(llvm_type_of(arg_type.element_of()),
+                                              lanes, scalable);
+        } else {
+            llvm_type = llvm_type_of(arg_type);
+        }
+        llvm_arg_types.push_back(llvm_type);
+    }
+    if (intrin.flags & RISCVIntrinsic::AddVLArg) {
+        mangled_name += (target.bits == 64) ? ".i64" : ".i32";
+        llvm_arg_types.push_back(xlen_type);
+    }
+
+    llvm::Function *inner =
+        get_llvm_intrin(llvm_ret_type, mangled_name, llvm_arg_types);
+    llvm::FunctionType *inner_ty = inner->getFunctionType();
+
+    // Remove vector tail preservation argument.
+    llvm_arg_types.erase(llvm_arg_types.begin());
+    // Remove vector length argument passed to inrinsic for wrapper.
+    // Wrapper will supply a constant for the fixed vector length.
+    if (intrin.flags & RISCVIntrinsic::AddVLArg) {
+        llvm_arg_types.resize(llvm_arg_types.size() - 1);
+    }
+
+    string wrapper_name = unique_name(std::string(intrin.name) + "_wrapper");
+    llvm::FunctionType *wrapper_ty = llvm::FunctionType::get(
+        inner_ty->getReturnType(), llvm_arg_types, false);
+    llvm::Function *wrapper =
+        llvm::Function::Create(wrapper_ty, llvm::GlobalValue::InternalLinkage,
+                               wrapper_name, module.get());
+    llvm::BasicBlock *block =
+        llvm::BasicBlock::Create(module->getContext(), "entry", wrapper);
+    llvm::IRBuilderBase::InsertPoint here = builder->saveIP();
+    builder->SetInsertPoint(block);
+
+    // Set vector fixed-point rounding flag if needed for intrinsic.
+    bool round_down = intrin.flags & RISCVIntrinsic::RoundDown;
```

Comment:
Do we need explicit `() != 0` here? The current flag values are fine, but if we ever get bits above the first 8, not sure if the language promises this to work :-/

Reply:
Conversion to ```bool``` is a comparison against zero. If the value is zero, it's false, otherwise true. Normally I would add the comparison explicitly, but for this usage, it feels idiomatic to treat the ```&``` operator as set inclusion. As in I feel adding the explicit test makes the code harder to read, not easier.